core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala

```diff
@@ -109,6 +111,9 @@ class KryoSerializer(conf: SparkConf)
     kryo.register(classOf[SerializableJobConf], new KryoJavaSerializer())
     kryo.register(classOf[HttpBroadcast[_]], new KryoJavaSerializer())
     kryo.register(classOf[PythonBroadcast], new KryoJavaSerializer())
+    kryo.register(classOf[TaskMetrics], new KryoJavaSerializer())
+    kryo.register(classOf[DirectTaskResult[_]], new KryoJavaSerializer())
+    kryo.register(classOf[IndirectTaskResult[_]], new KryoJavaSerializer())
```

Comment:
I just found there are more classes need to register, such as `ShuffleReadMetrics`, `OutputMetrics` and `ShuffleWriteMetrics`. The problem is that most fields in `TaskMetrics` are `Option` so the new added unit tests cannot cover them.

Not sure how many people uses `spark.kryo.registrationRequired`. But if they enable it, we will break their codes.

In addition, even if we register all necessary classes now, people may forget to register new classes if they just add an `Option` field in future.

I prefer to close this PR since I don't think this patch buys us much due to the tricky thing introduced.